File: backend/src/main/scala/customer/api/AddCustomerAddressApi.scala
package customer.api

import domain.shared.given

import cats.effect.IO
import customer.app.addCustomerAddress
import domain.customer.AddCustomerAddressRequest
import domain.shared.{CustomerId, DeliveryAppState, UserRole}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import shared.api.routing.*
import shared.app.*

val addCustomerAddressApi: FixedMethodApi1[CustomerId, DeliveryAppState] =
  jsonPostApi1[CustomerId, AddCustomerAddressRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("customers")),
    List(routeSegment("addresses")),
  )

File: frontend/src/apis/customer/AddCustomerAddressApi.ts
import type { AddCustomerAddressRequest, CustomerId, DeliveryAppState } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const addCustomerAddressApiDefinition = defineJsonPostApi1<CustomerId, AddCustomerAddressRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('customers')],
    [routeSegment('addresses')],
  )

export function addCustomerAddress(customerId: CustomerId, payload: AddCustomerAddressRequest) {
  return postNormalizedDeliveryState(buildApiPath1(addCustomerAddressApiDefinition, customerId), payload)
}
